### PR TITLE
DM-41063: Add nPixel filter for edge trailed sources and ignore filtering nans

### PR DESCRIPTION
**Pull request by bsmartradio**
Add several new flags to more accurately filter edge sources. Add new flag nan, off_image, and suspect_long_trail to be used in ap_association to more accurately filter sources which need to be removed.

### CODE REVIEW

File: python/lsst/meas/extensions/trailedSources/NaivePlugin.py [hunk 1]
```diff
@@ -124,7 +124,11 @@ def __init__(self, config, name, schema, metadata, logName=None):
         self.NO_CONVERGE = flagDefs.add("flag_noConverge", "The root finder did not converge")
         self.NO_SIGMA = flagDefs.add("flag_noSigma", "No PSF width (sigma)")
         self.SAFE_CENTROID = flagDefs.add("flag_safeCentroid", "Fell back to safe centroid extractor")
-        self.EDGE = flagDefs.add("flag_edge", "Trail contains edge pixels or extends off chip")
+        self.EDGE = flagDefs.add("flag_edge", "Trail contains edge pixels")
+        self.OFFIMAGE = flagDefs.add("flag_off_image", "Trail extends off image")
+        self.NAN = flagDefs.add("flag_nan", "One or more trail coordinates are missing")
+        self.SUSPECT_LONG_TRAIL = flagDefs.add("flag_suspect_long_trail",
+                                               "Trail length is greater than three times the psf radius")
```
Comment (mrawls) [inline python/lsst/meas/extensions/trailedSources/NaivePlugin.py:131]: These are good descriptions for `ap_association` where you need little "doc" strings for each flag!

File: python/lsst/meas/extensions/trailedSources/NaivePlugin.py [hunk 2]
```diff
@@ -282,6 +268,86 @@ def measure(self, measRecord, exposure):
         measRecord.set(self.keyLengthErr, lengthErr)
         measRecord.set(self.keyAngleErr, thetaErr)
 
+    def check_edge(self, measRecord, exposure, x0, y0, x1, y1, length):
+        """ Set flags for edge pixels, off chip, and nan trail coordinates.
```
Comment (mrawls) [inline python/lsst/meas/extensions/trailedSources/NaivePlugin.py]: Please note that this method also does the suspect_long_trail check with the PSF size.

File: python/lsst/meas/extensions/trailedSources/NaivePlugin.py [hunk 3]
```diff
@@ -282,6 +268,86 @@ def measure(self, measRecord, exposure):
...
+            y coordinate of the end of the trail.
+        """
+        x_coords = [x0, x1]
+        y_coords = [y0, y1]
+
+        # Check if one of the end points of the trail sources is nan. If so,
+        # set the trailed source nan flag.
+        if np.isnan(x_coords).any() or np.isnan(y_coords).any():
+            self.flagHandler.setValue(measRecord, self.NAN.number, True)
+
```
Comment (mrawls) [inline python/lsst/meas/extensions/trailedSources/NaivePlugin.py]: I feel like there are a lot of potentially-superfluous blank lines in this test, consider removing some unless they're really necessary for clarity.

File: python/lsst/meas/extensions/trailedSources/NaivePlugin.py [hunk 4]
```diff
@@ -282,6 +268,86 @@ def measure(self, measRecord, exposure):
...
+                # check any that have a nan.
+
+                if x0 is not np.nan and y0 is not np.nan:
+                    if exposure.mask[Point2I(int(x0), int(y0))] & exposure.mask.getPlaneBitMask('EDGE') != 0:
+                        self.flagHandler.setValue(measRecord, self.EDGE.number, True)
+
+                if x1 is not np.nan and y1 is not np.nan:
+                    if exposure.mask[Point2I(int(x1), int(y1))] & exposure.mask.getPlaneBitMask('EDGE') != 0:
+                        self.flagHandler.setValue(measRecord, self.EDGE.number, True)
+
```
Comment (mrawls) [inline python/lsst/meas/extensions/trailedSources/NaivePlugin.py]: Put this in a loop? I think `for (x_val, y_val) in zip(x_coords, y_coords):` would do both [x0, y0] and [x1, y1]

File: python/lsst/meas/extensions/trailedSources/NaivePlugin.py [hunk 5]
```diff
@@ -282,6 +268,86 @@ def measure(self, measRecord, exposure):
...
+        else:
+            # Check whether the beginning or end point of the trail has the
+            # edge flag set. The end points are not whole pixel values, so
+            # the pixel value must be rounded.
+            if exposure.mask[Point2I(int(x0), int(y0))] and exposure.mask[Point2I(int(x1), int(y1))]:
+                if ((exposure.mask[Point2I(int(x0), int(y0))] & exposure.mask.getPlaneBitMask('EDGE') != 0)
+                        or (exposure.mask[Point2I(int(x1), int(y1))]
+                            & exposure.mask.getPlaneBitMask('EDGE') != 0)):
+                    self.flagHandler.setValue(measRecord, self.EDGE.number,
+                                              True)
```
Comment (mrawls) [inline python/lsst/meas/extensions/trailedSources/NaivePlugin.py]: This dangling True feels exceptionally weirdly formatted. Perhaps run `black` after you run `isort` ...

File: tests/test_trailedEdgeSources.py [hunk 6]
```diff
@@ -231,8 +257,86 @@ def testEdgeFlag(self):
...
+            x0 = args[2]
+            y0 = args[3]
+            x1 = args[4]
+            y1 = np.nan  # overriden to test NAN flagging
+            length = args[6]
+            measRecord['ext_trailedSources_Naive_y1'] = np.nan
+            return original_check_edge_function(measRecord, exposure, x0, y0, x1, y1, length)
+
+        patcher = patch(
+            'lsst.meas.extensions.trailedSources.NaivePlugin.SingleFrameNaiveTrailPlugin.check_edge',
```
Comment (mrawls) [inline tests/test_trailedEdgeSources.py]: 😂

Comment (mrawls) [reply tests/test_trailedEdgeSources.py]: In all seriousness though, can you explain what you're mocking up here with a short comment, please?

**Review by mrawls — APPROVED** [review]

File: python/lsst/meas/extensions/trailedSources/NaivePlugin.py (hunk 2 above)
Comment (bsmartradio) [reply python/lsst/meas/extensions/trailedSources/NaivePlugin.py]: Renamed to `check_trail` as it now is checking more than the edge. Updated docstrings.

File: python/lsst/meas/extensions/trailedSources/NaivePlugin.py (hunk 3 above)
Comment (bsmartradio) [reply python/lsst/meas/extensions/trailedSources/NaivePlugin.py]: Removed the excess lines. They were mostly there for readability when I was testing.

File: python/lsst/meas/extensions/trailedSources/NaivePlugin.py (hunk 4 above)
Comment (bsmartradio) [reply python/lsst/meas/extensions/trailedSources/NaivePlugin.py]: Your right, changed into a loop.

File: python/lsst/meas/extensions/trailedSources/NaivePlugin.py (hunk 5 above)
Comment (bsmartradio) [reply python/lsst/meas/extensions/trailedSources/NaivePlugin.py]: My auto formatter got the best of me, thanks for catching it!

File: tests/test_trailedEdgeSources.py (hunk 6 above)
Comment (bsmartradio) [reply tests/test_trailedEdgeSources.py]: Added explanation for what the patcher is doing.

**Pull request merged**